pint_xarray/checking.py

```diff
@@ -0,0 +1,186 @@
+        and may optionally return one or more xarray.DataArrays or numpy-like arrays.
+    *args_units : unit-like or mapping of unit-like or None, optional
+        Units to expect for each positional argument given to func.
+
+        The decorator will first check that arguments passed to the decorated function possess these specific units
+        (or will attempt to convert the argument to these units), then will strip the units before passing the magnitude
+        to the wrapped function.
+
+        A value of None indicates not to check that argument for units (suitable for flags and other non-data
+        arguments).
+    return_units : unit-like or mapping of unit-like or list of unit-like or mapping of unit-like or None, optional
+        The expected units of the returned value(s), either as a single unit or as a list of units. The decorator
+        will attach these units to the variables returned from the function.
+
+        A value of None indicates not to attach any units to that return value (suitable for flags and other
+        non-data results).
+    kwargs_units : mapping of unit-like or None, Optional
+        Unit to expect for each keyword argument given to func.
+
+        The decorator will first check that arguments passed to the decorated function possess these specific units
+        (or will attempt to convert the argument to these units), then will strip the units before passing the magnitude
+        to the wrapped function.
+
+        A value of None indicates not to check that argument for units (suitable for flags and other non-data
+        arguments).
+
+    Returns
+    -------
+    return_values : Any
+        Return values of the wrapped function, either a single value or a tuple of values. These will be given units
+        according to return_units.
+
+    Raises
+    ------
+    TypeError
+        If an argument or return value has a specified unit, but is not an xarray.DataArray or pint.Quantity.
+        Also thrown if any of the units are not a valid type, or if the number of arguments or return values does not
+        match the number of units specified.
+
+    Examples
+    --------
+
+    Decorating a function which takes one quantified input, but returns a non-data value (in this case a boolean).
+
+    >>> @expects("deg C")
+    ... def above_freezing(temp):
+    ...     return temp > 0
+
+
+    TODO: example where we check units of an optional weighted kwarg
+    """
+
+    # Check types of args_units, kwargs_units, and return_units
+    all_units = list(args_units) + list(kwargs_units.values())
+    if isinstance(return_units, list):
+        all_units = all_units + return_units
+    elif return_units:
+        all_units = all_units + [return_units]
+    for a in all_units:
+        if isinstance(a, dict):
+            for u in a.values():
+                _check_valid_unit_type(u)
+        else:
+            _check_valid_unit_type(a)
```

Comment:
I think the reason I put this here was so that it would raise when the decorator was defined, rather than when the function was executed. Not sure if that's important though.

Reply:
@keewis thinking about this again - do you think there is any benefit to doing this check at decorator definition time? If so, then I think I should at least check if the unit exists in the registry, which is a lot more useful then simply checking if it is a string or Unit type.

Reply:
this would slow down the import time and speed up the call time (which is probably a good thing).

However, we can't really do that, at least for `str` typed units, because we need a registry for that (and for `Unit` instances we only need to check if the registry is the same). We could probably use the application registry, but then we'd be enforcing the use of the application registry.

So I guess we have to choose between checking in `.to` or in the `expects` wrapper function? I might be missing something, though.

Reply:
> However, we can't really do that, at least for str typed units, because we need a registry for that (and for Unit instances we only need to check if the registry is the same). We could probably use the application registry, but then we'd be enforcing the use of the application registry.

In that case it seems like that isn't much point in merely checking the type at import time, and I should try your suggestion of doing the check in `.to` instead, even though that would technically slow down the function call (you probably don't care about that slowdown if you're using pint at all to be fair).

Reply:
I changed it in https://github.com/xarray-contrib/pint-xarray/pull/143/commits/03503080a37a300bed387ef287397bb54c461cda